src/plot_api/plot_api.js

```diff
@@ -1576,9 +1576,17 @@ function _restyle(gd, aobj, _traces) {
                 helpers.manageArrayContainers(param, newVal, undoit);
                 flags.docalc = true;
             }
-            // all the other ones, just modify that one attribute
-            else param.set(newVal);
+            else {
+                var moduleAttrs = (contFull._module || {}).attributes || {};
+                var valObject = Lib.nestedProperty(moduleAttrs, ai).get() || {};
 
+                if(valObject.arrayOk && (Array.isArray(newVal) || Array.isArray(oldVal))) {
+                    flags.docalc = true;
```

Comment:
Previously, calls like `restyle(gd, 'marker.size', [[10, 30, 20]])` worked only when one or several axis `autorange` were turned on as this [too](https://github.com/plotly/plotly.js/blob/master/src/plot_api/plot_api.js#L1673) clear `gd.calcdata`.

This here now ensures that restyling any `arrayOk` key clears `gd.calcdata`.